### PR TITLE
minor memory improvements in _joint_likelihood() of ConditionalRandomField with advanced indexing

### PR DESCRIPTION
**Pull request by LauraRuis**
Minor memory improvements possible in the _joint_likelihood() function of the ConditionalRandomField class by utilizing the advanced indexing feature of PyTorch (https://github.com/pytorch/pytorch/releases/tag/v0.2.0). 

Elliminates the need to expand transitions matrix by batch size (line 258 - 260) and to expand the tensor containing indices to the last tags by batch size (line 289).

### CODE REVIEW

File: allennlp/modules/conditional_random_field.py [hunk 1]
```diff
@@ -255,26 +255,13 @@ def _joint_likelihood(self,
         else:
             score = 0.0
 
-        # Broadcast the transition scores to one per batch element
-        broadcast_transitions = self.transitions.view(1, num_tags, num_tags).expand(batch_size, num_tags, num_tags)
```
Comment (matt-gardner) [inline allennlp/modules/conditional_random_field.py]: You've removed the need for `num_tags`, so you need to remove the variable definition above on line 245 (`batch_size, sequence_length, _ = logits.size()`).

File: allennlp/modules/conditional_random_field.py [hunk 2]
```diff
@@ -286,10 +273,7 @@ def _joint_likelihood(self,
         # Transition from last state to "stop" state. To start with, we need to find the last tag
         # for each instance.
         last_tag_index = mask.sum(0).long() - 1
-        last_tags = tags.gather(0, last_tag_index.view(1, batch_size).expand(sequence_length, batch_size))
```
[the remark this answers is not in the document]
Comment (LauraRuis) [reply allennlp/modules/conditional_random_field.py]: Yes that makes more sense! I added your comments :)
